python/ray/serve/api.py

```diff
@@ -1451,3 +1517,348 @@ def get_deployment_statuses() -> Dict[str, DeploymentStatusInfo]:
     """
 
     return internal_get_global_client().get_deployment_statuses()
+
+
+class ImmutableDeploymentList(list):
+    def __init__(self, deployments: Iterable[Deployment]):
+        super().__init__()
+        [self.append(deployment) for deployment in deployments]
+
+    def __setitem__(self, *args):
+        """Not allowed. Modify deployment options using set_options instead."""
+        raise RuntimeError(
+            "Setting deployments in a built app is not allowed. Modify the "
+            'options using app.deployments["deployment"].set_options instead.'
+        )
+
+
+class Application:
+    """A static, pre-built Serve application.
+
+    An application consists of a number of Serve deployments that can send
+    requests to each other. One of the deployments acts as the "ingress,"
+    meaning that it receives external traffic and is the entrypoint to the
+    application.
+
+    The ingress deployment can be accessed via app.ingress and a dictionary of
+    all deployments can be accessed via app.deployments.
+
+    The config options of each deployment can be modified using set_options:
+    app.deployments["name"].set_options(...).
+
+    This application object can be written to a config file and later deployed
+    to production using the Serve CLI or REST API.
+    """
+
+    def __init__(self, deployments: List[Deployment], ingress: Deployment = None):
+
+        # TODO (shrekris-anyscale): validate ingress
+        self._ingress = ingress
+
+        deployments = deployments or []
+
+        self._deployments = dict()
+        for d in deployments:
+            self._add_deployment(d)
+
+    @property
+    def ingress(self) -> Deployment:
+        return self._ingress
+
+    @property
+    def deployments(self) -> ImmutableDeploymentList:
+        return ImmutableDeploymentList(self._deployments.values())
```

Comment:
hmm, why make this a list? we need to be able to access them by name to set config options, no?

Reply:
Ah good catch, let me change that.